### PR TITLE
[sled-agent] Avoid races on service import, properties on instances, explicit refresh

### LINKED ISSUE #1115: [sled-agent] Propolis server SMF service listen address is not always set correctly
I've noticed that occasionally a call to create/start a VM via the Oxide CLI seems to hang and then fail after about a minute. Inspecting the Nexus log, I see that the request timed out:

```
{"msg":"request completed","v":0,"name":"nexus","level":30,"time":"2022-05-25T05:32:37.432223873Z","hostname":"oxz_nexus","pid":20389,"uri":"//organizations/o/projects/p/instances/vm1/start","method":"POST","req_id":"85612f14-783f-40ea-982f-cc888d2320f8","remote_addr":"[fd00:1122:3344:101::1]:34100","local_addr":"[fd00:1122:3344:101::3]:12220","component":"dropshot_external","name":"e6bff1ff-24fb-49dc-a54e-c6a350cd4d6c","error_message_external":"Internal Server Error","error_message_internal":"CommunicationError: error sending request for url (http://[fd00:1122:3344:101::1]:12345/instances/8724453d-7959-44ed-809f-1bc9d02ebe9a): operation timed out","response_code":"500"}
```

Looking at the sled agent, I can see it hang after hitting [this line](https://github.com/oxidecomputer/omicron/blob/cb75b16e27755d86146833c2e5ffef9e07288a4d/sled-agent/src/instance.rs#L552):

```
{"msg":"Started propolis in zone: oxz_propolis-server_5d23c6ef-b4c0-4f0a-90de-fb5001133a5c","v":0,"name":"SledAgent","level":30,"time":"2022-05-25T05:31:45.446013495Z","hostname":"feldspar","pid":14194,"instance_id":"8724453d-7959-44ed-809f-1bc9d02ebe9a","component":"InstanceManager"}
{"msg":"halt_and_remove_logged: Previous zone state: Running","v":0,"name":"SledAgent","level":30,"time":"2022-05-25T05:32:39.406791793Z","hostname":"feldspar","pid":14194,"zone":"oxz_propolis-server_5d23c6ef-b4c0-4f0a-90de-fb5001133a5c","instance_id":"8724453d-7959-44ed-809f-1bc9d02ebe9a","component":"InstanceManager"}
{"msg":"Stopped and uninstalled zone","v":0,"name":"SledAgent","level":30,"time":"2022-05-25T05:32:39.408091477Z","hostname":"feldspar","pid":14194,"zone":"oxz_propolis-server_5d23c6ef-b4c0-4f0a-90de-fb5001133a5c","instance_id":"8724453d-7959-44ed-809f-1bc9d02ebe9a","component":"InstanceManager"}
```

It just sits there for about a minute, then removes the zone and cleans up. I was able to snag the Propolis logs during this hang, and we see this:

```
bnaecker@feldspar : ~/omicron $ pfexec zlogin $(zoneadm list | tail -1) 'tail +0 -f $(svcs -L propolis-server)'
[ May 24 22:31:45 Enabled. ]
[ May 24 22:31:45 Executing start method ("ctrun -l child -o noorphan,regent /opt/oxide/propolis-server/bin/propolis-server run /var/svc/manifest/site/propolis-server/config.toml unknown &"). ]
[ May 24 22:31:45 Method "start" exited with status 0. ]
error: Invalid value for '<PROPOLIS_IP:PORT>': invalid IP address syntax
[ May 24 22:31:45 Stopping because all processes in service exited. ]
[ May 24 22:31:45 Executing stop method (:kill). ]
[ May 24 22:31:45 Executing start method ("ctrun -l child -o noorphan,regent /opt/oxide/propolis-server/bin/propolis-server run /var/svc/manifest/site/propolis-server/config.toml unknown &"). ]
[ May 24 22:31:45 Method "start" exited with status 0. ]
error: Invalid value for '<PROPOLIS_IP:PORT>': invalid IP address syntax
[ May 24 22:31:45 Stopping because all processes in service exited. ]
[ May 24 22:31:45 Executing stop method (:kill). ]
[ May 24 22:31:45 Executing start method ("ctrun -l child -o noorphan,regent /opt/oxide/propolis-server/bin/propolis-server run /var/svc/manifest/site/propolis-server/config.toml unknown &"). ]
[ May 24 22:31:45 Method "start" exited with status 0. ]
error: Invalid value for '<PROPOLIS_IP:PORT>': invalid IP address syntax
[ May 24 22:31:45 Stopping because all processes in service exited. ]
[ May 24 22:31:45 Executing stop method (:kill). ]
[ May 24 22:31:45 Executing start method ("ctrun -l child -o noorphan,regent /opt/oxide/propolis-server/bin/propolis-server run /var/svc/manifest/site/propolis-server/config.toml unknown &"). ]
[ May 24 22:31:45 Method "start" exited with status 0. ]
error: Invalid value for '<PROPOLIS_IP:PORT>': invalid IP address syntax
[ May 24 22:31:45 Stopping because all processes in service exited. ]
[ May 24 22:31:45 Executing stop method (:kill). ]
[ May 24 22:31:45 Restarting too quickly, changing state to maintenance. ]
```

That shows that the `propolis-server` binary was started with a listening address of `"unknown"`, which is obviously wrong. That comes from the [default value of the `config/server_addr`](https://github.com/oxidecomputer/propolis/blob/a11ccd5bf001e0c690c67d117f8beb06b0bc6914/smf/propolis-server/manifest.xml#L23) SMF property. The sled agent is supposed to be setting that property [here](https://github.com/oxidecomputer/omicron/blob/cb75b16e27755d86146833c2e5ffef9e07288a4d/sled-agent/src/instance.rs#L526-L532) to the actual IP address that Nexus provided for that Propolis server. That doesn't seem to always happen.

I modified the sled agent to print the values of that property for both the generic `propolis-server` SMF service, and for the specific instance of that service that it adds. Those seem to always show:

```
{"msg":"Generic Propolis server SMF properties: config/server_addr  astring  [fd00:1122:3344:101::17]:12400\n","v":0,"name":"SledAgent","level":30,"time":"2022-05-25T05:31:45.331659446Z","hostname":"feldspar","pid":14194,"instance_id":"8724453d-7959-44ed-809f-1bc9d02ebe9a","component":"InstanceManager"}
{"msg":"Propolis server SMF instance properties: ","v":0,"name":"SledAgent","level":30,"time":"2022-05-25T05:31:45.408559664Z","hostname":"feldspar","pid":14194,"instance_id":"8724453d-7959-44ed-809f-1bc9d02ebe9a","component":"InstanceManager"}
```

That is, the "generic" SMF service `svc:/system/illumos/propolis-server` seems to get the provided value. The specific instance of the service, in this case `svc:/system/illumos/propolis-server:vm-5d23c6ef-b4c0-4f0a-90de-fb5001133a5c` has an empty string as its property value. Presumably this is because the instance is supposed to inherit properties from its parent configuration, unless they've been overridden. Unfortunately, those are always the values, when the instance provision succeeds or hangs.

However, when it hangs, I've also seen that inspecting the property value with `svccfg` directly from the global zone shows that it remains unmodified:

```
bnaecker@feldspar : ~/omicron $ pfexec zlogin $(zoneadm list | tail -1) 'svccfg -v -s propolis-server listprop config/server_addr'
config/server_addr  astring  unknown
```

I really don't have much else yet, but it's getting late and I wanted to record my debugging state. I've experimented a bit with setting the property directly on the instance, but that fails with the error: `svccfg: No such property group "config"`. I've also tried calling `svcprop refresh`, but that seems to have no effect.

Overall, it appears there's a race or some other inconsistency in how the actual SMF service _instance_ we're launching gets the value of its property `config/server_addr`. My current steps to repro this are, unfortunately, just starting/stopping an instance a ton of times. I can usually hit this after a median of about 5 retries, but it's extremely variable.

### PR DESCRIPTION
**Pull request by smklein**
Fixes #1115 , hopefully.

## Previously

I used the following script:

```bash
set -x
oxide org     create myorg -D "description"
oxide project create --organization myorg myproject -D "description"
set +x

set -x
for i in {0..10000}; do
  INSTANCE_NAME="myinstance${i}"
  oxide instance create --organization myorg --project myproject ${INSTANCE_NAME} -D "description" -c 2 -m 1073741824 --hostname myinstance
  oxide instance stop   --confirm --organization myorg --project myproject ${INSTANCE_NAME}
  oxide instance delete --confirm --organization myorg --project myproject ${INSTANCE_NAME}
done
set +x
```

After < 10 iterations, I'd typically see an SMF failure, either due to a missing property, or due to some issue on import.

After discussion with @bnaecker and @rmustacc , we confirmed that the Propolis Server zone automatically imports the manifest, and that it may be racing with an explicit import.

## This PR

This PR performs the following actions:

1. Rather than importing the manifest, we simply try to add a service instance in a retry loop. As seen in the log files, this typically fails for the first couple invocations (with the output `svccfg: Pattern 'svc:/system/illumos/propolis-
server' doesn't match any instances or services`), but succeeds once the built-in `configd` service has finished importing the propolis manifest.
2. It sets configuration parameters on the service *instance*, rather than the service itself.
3. It explicitly refreshes the instance before enabling it, to make sure the properties are up-to-date.

The aforementioned "create an instance, stop it, destroy it" script has been running on my machine for nearly 100 iterations without a single failure using this PR, which is a major improvement.

### CODE REVIEW

File: sled-agent/src/instance.rs [hunk 1]
```diff
@@ -516,37 +518,80 @@ impl Instance {
...
 
+        info!(inner.log, "Refreshing instance");
         running_zone.run_cmd(&[
             crate::illumos::zone::SVCCFG,
             "-s",
-            "svc:/system/illumos/propolis-server",
-            "add",
-            &format!("vm-{}", inner.propolis_id()),
+            &smf_instance_name,
+            "refresh",
```
Comment (bnaecker) [inline sled-agent/src/instance.rs:586]: We had talked about putting an additional check here, that the `config/server_addr` property has actually taken for the instance. Do you think that's overly paranoid?

File: sled-agent/src/instance.rs [hunk 2]
```diff
@@ -516,37 +518,80 @@ impl Instance {
...
+                    "Failed to add {} as a service (retrying in {:?}): {}",
+                    instance_name,
+                    delay,
+                    err.to_string()
+                );
+            },
+        )
+        .await?;
+
+        info!(inner.log, "Adding service property group");
```
Comment (bnaecker) [inline sled-agent/src/instance.rs:562]: Maybe add the service name to these log messages? The actual address might also be nice, in the `svccfg setprop` call below.

**Review by bnaecker — APPROVED** [review]
Thanks for taking this on! Looks good, and sounds like you've not hit any similar races after a while through the instance creation loop. I have a couple of comments, but otherwise 👍 .

**Comment by bnaecker** [comment]
For reference, this _should_ close #1115.

File: sled-agent/src/instance.rs (hunk 1 above)
Comment (smklein) [reply sled-agent/src/instance.rs:586]: I think I'm okay omitting it, based on the data collected (500+ iterations without the server_addr missing), but I'd be happy to add it later if we see any issues?

File: sled-agent/src/instance.rs (hunk 2 above)
Comment (smklein) [reply sled-agent/src/instance.rs:562]: I'll add more context. These all have the instance ID as part of the context already, so the log messages look like:

```
[2022-05-27T00:08:00.58389195Z]  INFO: SledAgent/InstanceManager/6692 on thelio: Adding service property group (instance_id=cd7cbd04-5507-4c53-a1e0-1f1e6016e7bb)
[2022-05-27T00:08:00.617921961Z]  INFO: SledAgent/InstanceManager/6692 on thelio: Setting server address property (instance_id=cd7cbd04-5507-4c53-a1e0-1f1e6016e7bb)
[2022-05-27T00:08:00.649773537Z]  INFO: SledAgent/InstanceManager/6692 on thelio: Refreshing instance (instance_id=cd7cbd04-5507-4c53-a1e0-1f1e6016e7bb)
[2022-05-27T00:08:00.675333382Z]  INFO: SledAgent/InstanceManager/6692 on thelio: Enabling instance (instance_id=cd7cbd04-5507-4c53-a1e0-1f1e6016e7bb)   
```

This should make it easy to disambiguate if multiple instances are being allocated simultaneously

**Pull request merged**